app/backends/workflow-orchestration.ts

```diff
@@ -526,138 +537,162 @@ export async function handleWorkItemUpdate(
+          tx, workItem.jobID, workItem.workflowStepIndex + 1,
+        );
+        if (nextWorkflowStep && (status !== WorkItemStatus.FAILED || nextWorkflowStep?.isBatched)) {
+          createdWorkItem = await createNextWorkItems(
+            tx,
+            nextWorkflowStep,
+            logger,
+            workItem,
+            status,
+            allWorkItemsForStepComplete,
+            results,
+            outputItemSizes,
+          );
         }
-        // If all granules are finished mark the job as finished
-        job.completeBatch(thisStep.workItemCount);
-        if (allWorkItemsForStepComplete) {
-          const finalStatus = await getFinalStatusForJob(tx, job);
-          await completeJob(tx, job, finalStatus, logger);
-        } else {
-          // Special case to pause the job as soon as any single granule completes when in the previewing state
-          if (job.status === JobStatus.PREVIEWING) {
-            job.pause();
+
+        if (nextWorkflowStep && status === WorkItemStatus.SUCCESSFUL) {
+          if (results && results.length > 0) {
+            // set the scrollID for the next work item to the one we received from the update
+            workItem.scrollID = scrollID;
+            await maybeQueueQueryCmrWorkItem(tx, workItem, logger);
+          } else {
+            // Failed to create the next work items when there should be work items.
+            // Fail the job rather than leaving it orphaned in the running state
+            logger.error('The work item update should have contained results to queue a next work item, but it did not.');
+            const message = 'Harmony internal failure: could not create the next work items for the request.';
+            await completeJob(tx, job, JobStatus.FAILED, logger, message);
+          }
+        } else if (!nextWorkflowStep || allWorkItemsForStepComplete) {
+          // Finished with the chain for this granule
+          if (status != WorkItemStatus.FAILED) {
+            await addJobLinksForFinishedWorkItem(tx, job, results, logger);
+          }
+          // If all granules are finished mark the job as finished
+          job.completeBatch(thisStep.workItemCount);
+          if (allWorkItemsForStepComplete && !createdWorkItem && (!nextWorkflowStep || nextWorkflowStep.workItemCount === 0)) {
+            const finalStatus = await getFinalStatusForJob(tx, job);
+            await completeJob(tx, job, finalStatus, logger);
+          } else {
+            // Either previewing or next step is a batched step and this item failed
+            if (job.status === JobStatus.PREVIEWING) {
+              // Special case to pause the job as soon as any single granule completes when in the previewing state
+              job.pause();
+            }
+            await job.save(tx);
           }
+        } else { // Currently only reach this condition for batched aggregation requests
           await job.save(tx);
```

Comment:
We are calling `job.save(tx)` from several places. I wonder if we should pull it of of the `if/else` blocks and do it afterwards. We could rewrite `completeJob` to not `save` the job to avoid the redundant save.